File: canonicalizer-go/internal/canonical/merkle.go
package canonical

import (
	"fmt"
	"math/big"

	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
)

const (
	rowBytes     = 548
	packChunk    = 31
	packedCount  = 18 // ceil(548/31)
	TreeDepth    = 17
	TreeCapacity = 1 << TreeDepth
)

// Context 把冻结的 schema 两半 + 域标签 + poseidon 实例绑在一起，使每个叶子/节点
// 计算都用相同的输入。
type Context struct {
	P          *Poseidon
	SchemaHi   fr.Element
	SchemaLo   fr.Element
	TagRow     fr.Element
	TagPadding fr.Element
	TagNode    fr.Element
}

// PackRowFields 把 548 字节 blob 拆成 18 个域元素（31 字节小端）。
func PackRowFields(blob []byte) ([]fr.Element, error) {
	if len(blob) != rowBytes {
		return nil, fmt.Errorf("expected %d-byte row", rowBytes)
	}
	out := make([]fr.Element, 0, packedCount)
	for offset := 0; offset < len(blob); offset += packChunk {
		end := offset + packChunk
		if end > len(blob) {
			end = len(blob)
		}
		chunk := blob[offset:end]
		rev := make([]byte, len(chunk)) // 反转：小端 → 大端，供 SetBytes 使用
		for i := range chunk {
			rev[i] = chunk[len(chunk)-1-i]
		}
		var e fr.Element
		e.SetBigInt(new(big.Int).SetBytes(rev))
		out = append(out, e)
	}
	return out, nil
}

// RowLeaf = H_P(TAG_ROW, [schemaHi, schemaLo, index, packed...])
func (c *Context) RowLeaf(index uint64, blob []byte) (fr.Element, error) {
	packed, err := PackRowFields(blob)
	if err != nil {
		return fr.Element{}, err
	}
	elements := make([]fr.Element, 0, 3+len(packed))
	elements = append(elements, c.SchemaHi, c.SchemaLo)
	var idx fr.Element
	idx.SetUint64(index)
	elements = append(elements, idx)
	elements = append(elements, packed...)
	return c.P.HashPoseidon(c.TagRow, elements), nil
}

// PaddingLeaf = H_P(TAG_PADDING, [schemaHi, schemaLo, index])
func (c *Context) PaddingLeaf(index uint64) fr.Element {
	var idx fr.Element
	idx.SetUint64(index)
	return c.P.HashPoseidon(c.TagPadding,
		[]fr.Element{c.SchemaHi, c.SchemaLo, idx})
}

// NodeHash = H_P(TAG_NODE, [level, left, right])
func (c *Context) NodeHash(level int, left, right fr.Element) fr.Element {
	var lvl fr.Element
	lvl.SetUint64(uint64(level))
	return c.P.HashPoseidon(c.TagNode, []fr.Element{lvl, left, right})
}

// BuildRoot 为 blobs（padding 到 capacity 叶子）构建 Merkle 根。
func (c *Context) BuildRoot(blobs [][]byte, capacity int) (fr.Element, error) {
	if len(blobs) > capacity {
		return fr.Element{}, fmt.Errorf("row count %d exceeds capacity %d", len(blobs), capacity)
	}
	leaves := make([]fr.Element, capacity)
	for i, b := range blobs {
		leaf, err := c.RowLeaf(uint64(i), b)
		if err != nil {
			return fr.Element{}, err
		}
		leaves[i] = leaf
	}
	for i := len(blobs); i < capacity; i++ {
		leaves[i] = c.PaddingLeaf(uint64(i))
	}

	level := 0
	for len(leaves) > 1 {
		next := make([]fr.Element, len(leaves)/2)
		for i := 0; i < len(leaves); i += 2 {
			next[i/2] = c.NodeHash(level, leaves[i], leaves[i+1])
		}
		leaves = next
		level++
	}
	return leaves[0], nil
}

// GeneratedFeatures 重现 synthetic-v1 的确定性特征向量。
// 必须与 Python experiments/g1/generate_vectors.generated_features 一致。
func GeneratedFeatures(i int) [FeatureCount]int32 {
	var out [FeatureCount]int32
	for j := 0; j < FeatureCount; j++ {
		out[j] = int32(((i + 1) * (j + 1)) & 0xFFFF)
	}
	return out
}

// EncodeGeneratedRow 为 synthetic-v1 的第 i 行构造已编码 blob。
func EncodeGeneratedRow(i int) ([]byte, error) {
	row := &Row{
		RowID:       uint64(i),
		Timestamp:   uint64(1700000000 + i),
		Features:    GeneratedFeatures(i),
		MissingMask: [MaskSize]byte{},
		Label:       1,
		Valid:       1,
	}
	if i%2 != 0 {
		row.Label = -1
	}
	return row.Encode()
}

// HexElement 把域元素渲染为 0x 前缀的小写规范十六进制。
func HexElement(e fr.Element) string {
	bi := e.BigInt(new(big.Int))
	return "0x" + bi.Text(16)
}

File: experiments/g1/merkle.py
"""Merkle tree over canonical rows using the Poseidon2 sponge.

A row's 548-byte blob is packed into 18 field elements (31-byte little-endian
chunks). The row leaf binds the schema hash, the row index and the packed
elements; padding leaves bind only the schema hash and index; internal nodes
bind the level and the two children. The tree has depth 17 (capacity 2^17).

Large trees (full 131072 capacity) are built with a process pool: the leaf
layer and every node layer are independent within a layer, so they parallelise
cleanly across CPU cores.
"""
from __future__ import annotations

import os
from concurrent.futures import ProcessPoolExecutor
from typing import Sequence

from .poseidon import hash_poseidon

ROW_BYTES = 548
PACK_CHUNK = 31
PACKED_ELEMENT_COUNT = 18  # ceil(548 / 31)
TREE_DEPTH = 17
TREE_CAPACITY = 1 << TREE_DEPTH

# Above this many leaves we spread the (embarrassingly parallel) per-layer work
# across worker processes.
_PARALLEL_THRESHOLD = 4096


def pack_row_fields(blob: bytes) -> list[int]:
    """Pack a 548-byte row into 18 field elements (31-byte LE chunks)."""
    if len(blob) != ROW_BYTES:
        raise ValueError("Expected 548-byte row")
    return [
        int.from_bytes(blob[offset : offset + PACK_CHUNK], "little")
        for offset in range(0, len(blob), PACK_CHUNK)
    ]


def row_leaf(
    index: int,
    blob: bytes,
    schema_hi: int,
    schema_lo: int,
    tag_row: int,
) -> int:
    """Data-row leaf: H_P(TAG_ROW, [schemaHi, schemaLo, index, r0..r17])."""
    elements = [schema_hi, schema_lo, index, *pack_row_fields(blob)]
    return hash_poseidon(tag_row, elements)


def padding_leaf(
    index: int,
    schema_hi: int,
    schema_lo: int,
    tag_padding: int,
) -> int:
    """Padding leaf: H_P(TAG_PADDING, [schemaHi, schemaLo, index])."""
    return hash_poseidon(tag_padding, [schema_hi, schema_lo, index])


def node_hash(
    level: int,
    left: int,
    right: int,
    tag_node: int,
) -> int:
    """Internal node: H_P(TAG_NODE, [level, left, right])."""
    return hash_poseidon(tag_node, [level, left, right])


# --- Worker functions (module-level so they pickle for the process pool). ---


def _leaf_chunk_worker(args):
    """Compute leaves [lo, hi) for the data/padding boundary at ``data_count``."""
    lo, hi, data_count, blobs, schema_hi, schema_lo, tag_row, tag_padding = args
    out = []
    for index in range(lo, hi):
        if index < data_count:
            out.append(row_leaf(index, blobs[index], schema_hi, schema_lo, tag_row))
        else:
            out.append(padding_leaf(index, schema_hi, schema_lo, tag_padding))
    return out


def _node_chunk_worker(args):
    """Hash the (left, right) pairs in one slice of a node layer."""
    level, slice_pairs, tag_node = args
    return [
        node_hash(level, left, right, tag_node) for left, right in slice_pairs
    ]


def _chunks(total: int, chunk: int):
    """Yield (lo, hi) bounds splitting [0, total) into ``chunk``-sized pieces."""
    for lo in range(0, total, chunk):
        yield lo, min(lo + chunk, total)


def build_root(
    row_blobs: Sequence[bytes],
    schema_hi: int,
    schema_lo: int,
    tag_row: int,
    tag_padding: int,
    tag_node: int,
    capacity: int = TREE_CAPACITY,
) -> int:
    """Build the data root for ``row_blobs`` padded up to ``capacity`` leaves."""
    if len(row_blobs) > capacity:
        raise ValueError(f"row count {len(row_blobs)} exceeds capacity {capacity}")

    row_blobs = list(row_blobs)
    use_parallel = capacity >= _PARALLEL_THRESHOLD and os.cpu_count() and os.cpu_count() > 1

    # Layer 0: data leaves then padding leaves.
    if use_parallel:
        leaves = _parallel_leaves(
            row_blobs, capacity, schema_hi, schema_lo, tag_row, tag_padding
        )
    else:
        leaves = [
            row_leaf(i, row_blobs[i], schema_hi, schema_lo, tag_row)
            for i in range(len(row_blobs))
        ]
        leaves.extend(
            padding_leaf(i, schema_hi, schema_lo, tag_padding)
            for i in range(len(row_blobs), capacity)
        )

    # Reduce pairwise up to the root; level participates in every node hash.
    level = 0
    while len(leaves) > 1:
        pairs = list(zip(leaves[0::2], leaves[1::2]))
        if use_parallel and len(pairs) >= _PARALLEL_THRESHOLD:
            leaves = _parallel_nodes(level, pairs, tag_node)
        else:
            leaves = [node_hash(level, l, r, tag_node) for l, r in pairs]
        level += 1

    return leaves[0]


def _parallel_leaves(blobs, capacity, schema_hi, schema_lo, tag_row, tag_padding):
    """Spread the leaf layer across worker processes."""
    workers = min(os.cpu_count() or 1, 8)
    chunk = max(1, capacity // (workers * 4))
    jobs = [
        (lo, hi, len(blobs), blobs, schema_hi, schema_lo, tag_row, tag_padding)
        for lo, hi in _chunks(capacity, chunk)
    ]
    leaves = []
    with ProcessPoolExecutor(max_workers=workers) as pool:
        for part in pool.map(_leaf_chunk_worker, jobs):
            leaves.extend(part)
    return leaves


def _parallel_nodes(level, pairs, tag_node):
    """Spread one node layer across worker processes."""
    workers = min(os.cpu_count() or 1, 8)
    chunk = max(1, len(pairs) // (workers * 4))
    jobs = [
        (level, pairs[lo:hi], tag_node) for lo, hi in _chunks(len(pairs), chunk)
    ]
    out = []
    with ProcessPoolExecutor(max_workers=workers) as pool:
        for part in pool.map(_node_chunk_worker, jobs):
            out.extend(part)
    return out

